src/main/kotlin/com/mdrobnak/lalrpop/psi/util/LpPsiUtil.kt

```diff
@@ -11,12 +18,22 @@ val List<LpSymbol>.selected: List<LpSymbol>
         this
     }
 
-fun List<LpSymbol>.computeType(arguments: List<NonterminalGenericArgument>): String {
+fun List<LpSymbol>.computeType(context: LpTypeResolutionContext, arguments: List<NonterminalGenericArgument>): String {
     val sel = selected
     return sel.joinToString(
         prefix = if (sel.size != 1) "(" else "",
         postfix = if (sel.size != 1) ")" else ""
     ) {
-        it.resolveType(arguments)
+        it.resolveType(context, arguments)
     }
 }
+
+fun PsiFile.lalrpopTypeResolutionContext(): LpTypeResolutionContext {
+    val externTokens = PsiTreeUtil.findChildrenOfType(this, LpExternToken::class.java)
+
+    val locationType = externTokens.mapNotNull { it.resolveLocationType() }.firstOrNull() ?: "usize"
+    val errorType = externTokens.mapNotNull { it.resolveErrorType() }.firstOrNull() ?: "()"
+    val tokenType = externTokens.mapNotNull { it.resolveTokenType() }.firstOrNull() ?: "&str"
+
+    return LpTypeResolutionContext(locationType = locationType, errorType = errorType, tokenType = tokenType)
```

Comment:
nit: this could be less verbose.
```suggestion
    return LpTypeResolutionContext(locationType, errorType, tokenType)
```